### PR TITLE
Components support for cmake_find_package_multi generator

### LINKED ISSUE #6718: [feature] Components: POC components in the cmake generator
After #6716 and #6717, we should be able to do a POC with the CMake generator and the usage of the "smart targets": Targets that depend only on the components needed (not the whole library. 

With the possibility of linking just with the needed components of other requirements depending on the implementation of #6717

### PR DESCRIPTION
**Pull request by danimtb**
Changelog: Feature: Support components for `cmake_find_package_multi` generator.
Docs: https://github.com/conan-io/docs/pull/1755

#TAGS: slow

- [x] Refer to the issue that supports this Pull Request: closes #6718 
- [x] If the issue has missing info, explain the purpose/use case/pain/need that covers this Pull Request.
- [x] I've read the [Contributing guide](https://github.com/conan-io/conan/blob/develop/.github/CONTRIBUTING.md).
- [x] I've followed the PEP8 style guides for Python code.
- [x] I've opened another PR in the Conan docs repo to the ``develop`` branch, documenting this one. 

<sup>**Note:** By default this PR will skip the slower tests and will use a limited set of python versions. Check [here](https://github.com/conan-io/conan/blob/develop/.github/PR_INCREASE_TESTING.md) how to increase the testing level by writing some tags in the current PR body text.</sup>

### CODE REVIEW

File: conans/client/generators/cmake_find_package_multi.py [hunk 1]
```diff
@@ -91,33 +94,273 @@ class CMakeFindPackageMultiGenerator(Generator):
...
+                         $<$<CONFIG:Release>:{{ '${'+pkg_name+'_'+comp_name+'_COMPILE_DEFINITIONS_RELEASE}' }}>
+                         $<$<CONFIG:RelWithDebInfo>:{{ '${'+pkg_name+'_'+comp_name+'_COMPILE_DEFINITIONS_RELWITHDEBINFO}' }}>
+                         $<$<CONFIG:MinSizeRel>:{{ '${'+pkg_name+'_'+comp_name+'_COMPILE_DEFINITIONS_MINSIZEREL}' }}>
+                         $<$<CONFIG:Debug>:{{ '${'+pkg_name+'_'+comp_name+'_COMPILE_DEFINITIONS_DEBUG}' }}>)
+        set_property(TARGET {{ pkg_name }}::{{ comp_name }} PROPERTY INTERFACE_COMPILE_OPTIONS
+                         $<$<CONFIG:Release>:{{ '${'+pkg_name+'_'+comp_name+'_COMPILE_OPTIONS_LIST_RELEASE}' }}>
+                         $<$<CONFIG:RelWithDebInfo>:{{ '${'+pkg_name+'_'+comp_name+'_COMPILE_OPTIONS_LIST_RELWITHDEBINFO}' }}>
+                         $<$<CONFIG:MinSizeRel>:{{ '${'+pkg_name+'_'+comp_name+'_COMPILE_OPTIONS_LIST_MINSIZEREL}' }}>
+                         $<$<CONFIG:Debug>:{{ '${'+pkg_name+'_'+comp_name+'_COMPILE_OPTIONS_LIST_DEBUG}' }}>)
+        set({{ pkg_name }}_{{ comp_name }}_TARGET_PROPERTIES TRUE)
```
Comment (danimtb) [inline conans/client/generators/cmake_find_package_multi.py:246]: This is a trick to "replace" the global target when a component with the global name already exists

File: conans/client/generators/cmake_find_package_multi.py [hunk 2]
```diff
@@ -91,33 +94,273 @@ class CMakeFindPackageMultiGenerator(Generator):
...
+        find_package_components = []
+        for comp_name, comp in self.sorted_components(cpp_info).items():
+            comp_findname = self._get_name(cpp_info.components[comp_name])
+            deps_cpp_cmake = DepsCppCmake(comp)
+            deps_cpp_cmake.public_deps = self._get_component_requires(pkg_name, pkg_findname, comp)
+            find_package_components.append((comp_findname, deps_cpp_cmake))
+        find_package_components.reverse()  # From the less dependent to most one
+        return find_package_components
+
+    def _get_component_requires(self, pkg_name, pkg_findname, comp):
```
Comment (danimtb) [inline conans/client/generators/cmake_find_package_multi.py]: These two methods are copied from the `cmake_find_package` generator class

**Comment by danimtb** [comment]
I see there are some tests failing only on Mac, so I can't reproduce the behavior from my end 😕

**Comment by jgsogo** [comment]
Running locally in Mac, this is the horror picture: not deterministic, probably race conditions with the underlying filesystem.

Having a look to failing test `basic_test`, and inspecting the output, I can see differences between runs (these are outputs from the `conan build fake_test_package` calls for the `Debug` build):

```
----Running------
> cmake --build '/private/var/folders/fc/6mvcrc952dqcjfhl4c7c11ph0000gn/T/tmp8ml_ubcfconans/path with spaces' '--' '-j1'
-----------------
Scanning dependencies of target example
[100%] Built target example
``` 

```
----Running------
> cmake --build '/private/var/folders/fc/6mvcrc952dqcjfhl4c7c11ph0000gn/T/tmp7c0000jyconans/path with spaces' '--' '-j1'
-----------------
Scanning dependencies of target example
[ 50%] Building CXX object CMakeFiles/example.dir/example.cpp.o
[100%] Linking CXX executable bin/example
[100%] Built target example
```

Adding a sleep before running the `Debug` section always works... you can try to generate the `example` in a different folder based on the build_type.

**Review by memsharded — APPROVED** [review]

**Review by jgsogo — APPROVED** [review]

**Pull request merged**